cachi2/core/package_managers/pip.py

```diff
@@ -1386,14 +1390,23 @@ def _download_dependencies(
     pip_deps_dir = output_dir.join_within_root("deps", "pip")
     pip_deps_dir.path.mkdir(parents=True, exist_ok=True)
 
-    downloads = []
+    downloaded = []
+    to_download: list[DistributionPackageInfo] = []
 
     for req in requirements_file.requirements:
         log.info("Downloading %s", req.download_line)
 
         if req.kind == "pypi":
-            download_info = _download_pypi_package(req, pip_deps_dir, PYPI_URL)
-            _check_metadata_in_sdist(download_info["path"])
+            source, wheels = _process_package_distributions(
+                req, pip_deps_dir, user_checksums, allow_binary
+            )
+            if allow_binary:
+                to_download.extend([w for w in wheels if not w.path.exists()])
```

Comment:
yeah, "existing output folder" is not something that we handle at all

at some point, it might be good to reach a state where we re-use and verify the content of an existing output dir, but that's a whole new feature

Reply:
OK, I removed extra if statement...
`to_download.extend(wheels)`

Reply:
I think that breaks Scenario 2 as we're discussing [here](https://github.com/containerbuildsystem/cachi2/pull/255#discussion_r1301180319), no?

If you always {download -> verify hash -> delete}, then you end up keeping only the wheels which match the hashes in the last requirements file that you process

Reply:
So if the file already exists but does not matches any of the provided hashes, should we keep it the file in the output directory ?

I put the line there based on this sentence in design doc:
_If the file already exists in the deps/pip directory, don’t re-download it. For atomic-reactor, we could save 63 downloads (out of 343)._

Reply:
We don't want to re-download the same wheels because we want to speed things up and save network bandwidth (that's from the design doc)

We don't want to re-check hashes for already downloaded wheels, because there's nothing necessarily wrong if a specific wheel matches in one requirements file and doesn't match in another one (I didn't realize this during the design, we figured this out in the PR)

Reply:
^ (the second point above assumes that we handle checksums on a single-file level)

Reply:
Do we want a ticket to track this? I thought that cleaning the folder or failing the build would be easy to implement choices. Currently, we overwrite all overlapping files there, so I think this would make the behavior at least obvious.

Reply:
why not :+1: